Labeling Sessions - Create DatabricksLabelingStore and AbstractLabelingStore (Refactor)

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/17931?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17931/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17931/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17931/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Labeling Sessions - Create DatabricksStore (Refactor)

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
store

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove backend session coupling from LabelingSession and fix error handling

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix label schema tests by mocking store selection properly

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix store mocking in label schema tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move _get_store to LabelingSession instance method to eliminate duplication

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor labeling stores with improved type hints and utilities

- Add comprehensive type hints to all methods in labeling stores module
- Fix traces parameter type from Any to Union[Iterable[Trace], Iterable[str], pd.DataFrame]
- Use TYPE_CHECKING imports for Databricks session types
- Rename _from_backend_session to _databricks_session_to_labeling_session for clarity
- Create databricks_utils.py to centralize Databricks import logic and error handling
- Eliminate 7 duplicate try/except blocks with single get_review_app() utility function
- Create reusable pytest fixtures for Databricks labeling store mocking
- Remove duplicate _ERROR_MSG definitions across modules
- All tests pass with improved code maintainability and type safety

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Rename get_review_app to get_databricks_review_app for clarity

Removes the need for import aliases and makes the function name more explicit about what it returns.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use mock fixture in test_api_integration_with_all_input_types

Replaces redundant manual mocking setup with the reusable mock_databricks_labeling_store fixture for cleaner and more consistent test code.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Clean up delete_label_schema to use get_databricks_review_app

- Replace duplicate try/except import logic with get_databricks_review_app utility
- Remove unused _ERROR_MSG definition in label_schemas/__init__.py
- All Databricks imports now consistently go through databricks_utils module

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/label_schemas/__init__.py:
```diff
@@ -93,22 +85,13 @@ def get_label_schema(name: str) -> LabelSchema:
 
 
-def delete_label_schema(name: str) -> "ReviewApp":
+def delete_label_schema(name: str):
```
The fact that `delete_label_schema` returns a `ReviewApp` instance is strange. I think we should preserve the behavior on Databricks (backcompat), but it seems more appropriate to return `None` in OSS.

Accordingly, I've omitted a type hint and return type docstring here. Users don't expect deletion to return a ReviewApp object, and None doesn't need to be documented.

This seems like a reasonably safe, backwards-compatible middle ground

dbczumar commented on mlflow/genai/labeling/__init__.py:
```diff
@@ -112,21 +94,11 @@ def get_labeling_session(run_id: str) -> LabelingSession:
 
 
-def delete_labeling_session(labeling_session: LabelingSession) -> "ReviewApp":
+def delete_labeling_session(labeling_session: LabelingSession):
```
Ditto - see https://github.com/mlflow/mlflow/pull/17931/files#r2377680983

dbczumar commented on mlflow/genai/labeling/stores.py:
```diff
@@ -0,0 +1,487 @@
+    def add_traces_to_session(
+        self,
+        labeling_session: LabelingSession,
+        traces: Union[Iterable[Trace], Iterable[str], "pd.DataFrame"],
```
As a follow-up, this can likely just be simplified to a set of Trace objects or trace IDs at the store level, but out of scope for this PR.

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update add_traces_to_session to handle DataFrame and string inputs

- Convert pandas DataFrame with 'trace' column to list of traces
- Handle string representations that need deserialization
- Use MlflowException for proper error handling
- Pass list of Trace objects to backend instead of raw input

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move trace conversion logic from store to LabelingSession

- Store now only accepts list[Trace], not mixed types
- LabelingSession.add_traces handles DataFrame and string conversion
- Use Trace.from_json() for deserializing JSON strings
- Cleaner separation of concerns between session and store

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary comments from trace conversion code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/labeling/labeling.py:
```diff
@@ -143,7 +175,39 @@ def add_traces(
             LabelingSession: The updated labeling session.
         """
-        return LabelingSession(self._session.add_traces(traces))
+        import pandas as pd
+
+        from mlflow.exceptions import MlflowException
+        from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+
+        if isinstance(traces, pd.DataFrame):
+            if "trace" not in traces.columns:
+                raise MlflowException(
+                    "traces must have a 'trace' column like the result of mlflow.search_traces()",
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
+            traces = traces["trace"].to_list()
+
+        trace_list: list[Trace] = []
+        for trace in traces:
+            if isinstance(trace, str):
+                trace_list.append(Trace.from_json(trace))
+            elif isinstance(trace, Trace):
+                trace_list.append(trace)
+            elif trace is None:
+                raise MlflowException(
+                    "trace cannot be None. Must be mlflow.entities.Trace or its json string "
+                    "representation.",
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
+            else:
+                raise MlflowException(
+                    f"Expected mlflow.entities.Trace or json string, got {type(trace).__name__}",
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
```
For reference, see the implementation of LabelingSession.add_traces in `databricks.agents`

dbczumar commented on mlflow/genai/labeling/labeling.py:
```diff
@@ -143,7 +175,39 @@ def add_traces(
+                    f"Expected mlflow.entities.Trace or json string, got {type(trace).__name__}",
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
+
+        store = self._get_store()
+        return store.add_traces_to_session(self, trace_list)
```
It's cleaner if the store always just takes `Trace` objects

dbczumar commented on mlflow/genai/labeling/labeling.py:
```diff
@@ -46,63 +43,97 @@ class LabelingSession:
 
-    def __init__(self, session: "_LabelingSession"):
-        self._session = session
+    def __init__(
+        self,
+        *,
+        name: str,
```
I checked and verified that there are no call sites for the `LabelingSession` constructor from a Databricks Labeling Session within MLflow or known projects. This seems like a safe change.

BenWilson2 reviewed (approved):
Abstractions and patterns look great! 
comments are entirely minor nits.
This certainly sets up the standard MLflow implementation to be straightforward!

dbczumar committed:
Address review comments from BenWilson2 on PR #17931

- Move MlflowException and INVALID_PARAMETER_VALUE imports to top-level in labeling.py
- Move tracking service utils import to top-level in stores.py
- Add documentation to DatabricksLabelingStore.__init__ explaining the no-op design pattern

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move tracking_uri constructor to AbstractLabelingStore

- Define the constructor pattern in the abstract base class where it belongs
- Remove redundant constructor from DatabricksLabelingStore
- All store implementations now inherit the standard constructor signature

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify AbstractLabelingStore constructor docstring

Remove confusing technical details about polymorphism from the docstring

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/labeling/stores.py:
```diff
@@ -0,0 +1,485 @@
+    This store delegates all labeling operations to the Databricks agents API.
+    """
+
+    def __init__(self, tracking_uri: str | None = None) -> None:
```
Moved this to abstract labeling store
